test/framework/src/main/java/org/elasticsearch/cluster/coordination/AbstractCoordinatorTestCase.java

```diff
@@ -723,18 +726,55 @@ void clearBlackholedConnections() {
                         nodeEnvironment = null;
                         BytesStreamOutput outStream = new BytesStreamOutput();
                         outStream.setVersion(Version.CURRENT);
-                        final MetaData updatedMetaData = adaptGlobalMetaData.apply(oldState.getLastAcceptedState().metaData());
-                        final ClusterState clusterState;
-                        if (updatedMetaData != oldState.getLastAcceptedState().metaData()) {
-                            clusterState = ClusterState.builder(oldState.getLastAcceptedState()).metaData(updatedMetaData).build();
+
+                        final long persistedCurrentTerm;
+
+                        if ((oldState.getLastAcceptedState().nodes().getLocalNode().isMasterNode() && newLocalNode.isMasterNode()) == false
```

Comment:
this is only exercised by the new test that you've added, right?  We're not switching roles o.w.

Reply:
why not reset state when restarting non-master node? why only case if we switch from master to non-master

Reply:
This condition is true if the node was master-ineligible _either_ before _or_ after the restart. In particular it's true when restarting a master-ineligible node without changing its roles. I think you're parsing it wrong? I added comments in 81e162d88c7acf87cf9d5ac9bd74fc5e49c4cd6a to clarify.

Reply:
I parsed it wrong indeed, too many brackets 🎱